### PR TITLE
CP-20787: Allocate VF for MxGPU VGPU types when starting VM

### PR DESCRIPTION
**Pull request by robhoes**
Most of these patches refactor the current code, to make things easier to implement the final bits, and should by themselves not change anything. The real changes are in the commits marked with the CP-number. Next step: fix the race condition that exists when choosing free VFs.

### CODE REVIEW

File: ocaml/xapi/vgpuops.ml [hunk 1]
```diff
@@ -69,23 +70,31 @@ let allocate_vgpu_to_gpu ~__context vgpu available_pgpus =
...
+        Ref.string_of vgpu.gpu_group_ref;
+        Ref.string_of vgpu.type_ref
+      ]))
+  | Some `PF ->
     raise (Api_errors.Server_error (Api_errors.vm_requires_gpu, [
         Ref.string_of vm;
         Ref.string_of vgpu.gpu_group_ref
       ]))
+
+let create_vgpu ~__context ~vm ~host vgpu =
```
Comment (thomassa) [inline ocaml/xapi/vgpuops.ml]: Having a `create_vgpu` function return `pgpu` looks weird on the face of it.
Right now I understand it and it makes sense, but for the sake of anyone trying to understand this code in future, let's have a comment saying what the function does and what it returns.

File: ocaml/xapi/vgpuops.ml [hunk 2]
```diff
@@ -34,135 +35,130 @@ let vgpu_of_vgpu ~__context vm_r vgpu =
...
+        Ref.string_of vgpu.type_ref
+      ]))
+  | Some `PF ->
     raise (Api_errors.Server_error (Api_errors.vm_requires_gpu, [
         Ref.string_of vm;
         Ref.string_of vgpu.gpu_group_ref
       ]))
-  | Some (pgpu, pci) ->
+
+let create_vgpu ~__context ~vm ~host vgpu =
```
Comment (thomassa) [inline ocaml/xapi/vgpuops.ml]: As discussed, this is just a wrapper function now (and a misleadingly named one at that), and its functionality can be incorporated into `allocate_vgpu_to_gpu` which is not called from anywhere else. The `vm` is implicit in the `vgpu`.

File: ocaml/xapi/vgpuops.ml [hunk 3]
```diff
@@ -34,135 +35,130 @@ let vgpu_of_vgpu ~__context vm_r vgpu =
...
+  let rec get retry =
+    match Pciops.get_free_virtual_function ~__context pf with
+    | Some vf -> vf
+    | None ->
+      if retry then begin
+        (* We may still need to load the driver... do that and try again *)
+        Xapi_pgpu.mxgpu_vf_setup ~__context;
+        get false
+      end else
+        (* This probably means that our capacitity checking went wrong! *)
```
Comment (thomassa) [inline ocaml/xapi/vgpuops.ml]: Typo where "capacity" was meant.

File: ocaml/xapi/vgpuops.ml (hunk 2 above)
Comment (robhoes) [inline ocaml/xapi/vgpuops.ml]: Actually, the `vm` was still needed for the error case.

**Comment by robhoes** [comment]
@thomassa Updated.

**Comment by thomassa** [comment]
I'm happy with this set of commits.
Please would you just update the merge-request title and first comment to reflect the current content, now that it includes not just refactoring but also a race-prone preliminary implementation of the MxGPU allocation?

**Pull request merged**